## Lab: SQL Practice

> Author: Matt Brems

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [2]:
#!conda install psycopg2

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Solving environment: done

## Package Plan ##

  environment location: /Library/anaconda3/envs/dsi

  added / updated specs:
    - psycopg2


The following NEW packages will be INSTALLED:

  libpq              pkgs/main/osx-64::libpq-11.2-h051b688_0
  psycopg2           pkgs/main/osx-64::psycopg2-2.8.4-py37ha12b0ac_0

The following packages will be DOWNGRADED:

  curl                                    7.69.1-ha441bb4_0 --> 7.67.0-ha441bb4_0
  krb5                                    1.17.1-hddcf347_0 --> 1.16.4-hddcf347_0
  libcurl                                 7.69.1-h051b688_0 --> 7.67.0-h051b688_0
  python                           3.7.7-hc70fcce_0_cpython --> 3.7.7-hfe9666f_0_cpython
  readline                                   8.0-h1de35cc_0 --> 7.0-h1de35cc_5


Proceed ([y]/n)? ^C

CondaS

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine



# Set up the engine to access the data.
engine = create_engine('postgres://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(store)
FROM stores
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: There are 1973 rows.

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM stores
ORDER BY name ASC
"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0      4986  218 Fuel Express & Chubby's Liquor            A   
1      3047          3047 Cub Foods / Iowa City            I   
2      3060               3060 Cub Foods / Ames            I   
3      3061         3061 Cub Foods / Sioux City            I   
4      4845                  3rd St Convenience            A   
...     ...                                 ...          ...   
1968   3992       Xo Food And Liquor / Waterloo            I   
1969   4066                       Xpress Liquor            I   
1970   4886                      Yr Dollar Stop            A   
1971   4908                Zapf's Pronto Market            A   
1972   4874                      Z's Quickbreak            A   

                                          store_address address_info  
0     68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1     855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2     3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3     1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None  
4     216, 3rd St Se\nCedar Rapids, IA 52401\n(41.97...         None  
...                                                 ...          ...  
1968  428 Franklin St\nWaterloo, IA 507030000\n(42.5...         None  
1969  1201 North Jefferson Ste 100\nIndianola, IA 50...         None  
1970  1500 1st Ave Ne\nCedar Rapids, IA 52402\n(41.9...         None  
1971  107 1st St Sw\nElkader, IA 52043\n(42.85350191...         None  
1972  201, E Washington St\nMount Pleasant, IA 52641...         None  

[1973 rows x 5 columns]

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor

"""

pd.read_sql_query(sql, engine).head(3)


# SELECT f.name AS food, r.name
# FROM foods f
# INNER JOIN restaurants r ON r.id = f.restaurant_id

date convenience_store  store county_number county category  \
0 2014-07-01              None   4509            85  Story  1032080   
1 2014-11-24              None   2502            77   Polk  1032080   
2 2014-06-04              None   2548            77   Polk  1032080   

    category_name vendor_no                     vendor    item  \
0  IMPORTED VODKA        14  Adamba Imports Int'l Inc.  904783   
1  IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   
2  IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   

                   description  pack  liter_size state_btl_cost btl_price  \
0                 Vesica Vodka    12         750          $5.99     $8.99   
1  Michael Godard "Mr. Smooth"     6         750         $19.00    $28.50   
2  Michael Godard "Mr. Smooth"     6         750         $19.00    $28.50   

   bottle_qty   total  
0          12  107.88  
1           4  114.00  
2           6  171.00

**Prompt 3 Answer**:      
  
Price: 8.99, Quantity: 12     
Price: 28.50, Qty: 4    
Price: 28.50, Qty: 6     


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sum(total)
FROM sales
WHERE date >= '2015-02-01'
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**: Total Sales after Feb 1 2015 (there are no other sales after Feb): 21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.vendor AS sale, p.vendor_name
FROM sales s
INNER JOIN products p ON p.vendor_name = s.vendor

"""

# sql = """
# SELECT r.name, AVG(f.calories)
# FROM foods f
# INNER JOIN restaurants r ON r.id = f.restaurant_id
# GROUP BY r.name
# ORDER BY r.name
# """

pd.read_sql_query(sql, engine)

**Prompt 5 Answer**: 

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Create a column called `Markup` that shows the difference between the `btl_price` and the `state_btl_cost`. Take a screenshot and paste it into your notebook.

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """

"""

pd.read_sql_query(sql, engine)

**Prompt 6 Answer**: 

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

> Note: To make this an easier question, you may take the simple average of percentage markups across all transactions for a store. You do **not** need to weight the average markup by the number of bottles purchased per transaction. If you want to practice your SQL skills, though, consider this a bonus challenge!

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """

"""

pd.read_sql_query(sql, engine)

**Prompt 7 Answer**: 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """

"""

pd.read_sql_query(sql, engine)

**Prompt 8 Answer**:

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """

"""

pd.read_sql_query(sql, engine)

**Prompt 9 Answer**:

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """

"""

pd.read_sql_query(sql, engine)

**Prompt 10 Answer**: